test.py

```diff
@@ -277,20 +270,120 @@ async def find_tests(options: argparse.Namespace) -> None:
                 suite = TestSuite.opt_create(f, options, mode)
                 await suite.add_test_list()
 
-    if not TestSuite.test_count():
-        if len(options.name):
-            print("Test {} not found".format(palette.path(options.name[0])))
-            sys.exit(1)
+
+def run_pytest(options: argparse.Namespace, run_id: int) -> tuple[int, list[SimpleNamespace]]:
+    failed_tests = []
+    temp_dir = pathlib.Path(options.tmpdir).absolute()
+    report_dir =  temp_dir / 'report'
+    junit_output_file = report_dir / f'pytest_cpp_{run_id}.xml'
+    files_to_run = []
+    test_names = []
+    for name in options.name:
+        file_name = name
+        if '::' in name:
+            file_name, _ = name.split('::', maxsplit=1)
+        if any((TOP_SRC_DIR / file_name).is_relative_to(x) for x in PYTEST_RUNNER_DIRECTORIES):
+            files_to_run.append(name)
+    if not options.name :
+        files_to_run = [str(directory) for directory in PYTEST_RUNNER_DIRECTORIES]
+    if not files_to_run:
+        logging.info(f'No boost found. Skipping pytest execution for boost tests.')
+        return 0, []
+    expression = ' or '.join(test_names)
+    if options.skip_patterns:
+        expression += f'{"and not ".join(options.skip_patterns)}'
+    modes = ' '.join(f'--mode={mode}' for mode in options.modes)
+    args = [
+        'pytest',
+        "-s",  # don't capture print() output inside pytest
+        '--color=yes',
+        modes,
+    ]
+    if options.list_tests:
+        args.extend(['--collect-only', '--quiet'])
+    else:
+        args.extend([
+            "--log-level=DEBUG",  # Capture logs
+            f'--junit-xml={junit_output_file}',
+            "-rf",
+            f'-n{int(options.jobs)}',
+            f'--tmpdir={temp_dir}',
+            f'--run_id={run_id}',
+            f'--maxfail={options.max_failures}',
+            f'--alluredir={report_dir / "allure"}',
+            '-v' if options.verbose else '-q',
+        ])
+    if options.gather_metrics:
+        args.append('--gather-metrics')
+    if len(expression) > 1:
+        args.extend(['-k', expression])
+    if not options.save_log_on_success:
+        args.append('--allure-no-capture')
+    if options.markers:
+        args.append(f"-m={options.markers}")
+    args.extend(files_to_run)
+
+    args = shlex.split(' '.join(args))
+    p = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, bufsize=1, universal_newlines=True)
+    try:
+        # Read output from pytest and print it to the console
+        if options.verbose:
+            for line in iter(p.stdout.readline, ''):
```

Comment:
`for line in p.stdout:` is enough.

Funny stuff :) 

This code snippet came from official Python docs for the `iter()` builtin (you can still see it for 3.6), but it already fixed for 3.7+: https://bugs.python.org/issue34764